### PR TITLE
three logos aka 早上好中国，现在我有冰淇淋

### PR DESCRIPTION
**Pull request by envyaims**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: solutions/bronze/cf-581D.mdx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main() {
+	int n = 3;
```
Comment (SansPapyrus683) [inline solutions/bronze/cf-581D.mdx:25]: make this a constant

File: solutions/bronze/cf-581D.mdx [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+		return 0;
+	}
+	// loop through all rotations of each rectangle
+	for (int rotate_mask = 0; rotate_mask < (1 << n); rotate_mask++) {
+		vector<string> grid(len, string(len, 'Z'));  // temporary grid
+		// 'Z' represents an empty space in the grid
+		int num_placed = 0;
+		for (int i = 0; i < len; i++) {
+			for (int j = 0; j < len; j++) {
+				if (grid[i][j] == 'Z') {
```
Comment (SansPapyrus683) [inline solutions/bronze/cf-581D.mdx:46]: space this thing out jesus christ

File: solutions/bronze/cf-581D.mdx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main() {
+	int n = 3;
+	vector<pair<int, int>> logos(n);
+	for (int i = 0; i < n; i++) {
+		scanf("%d%d", &logos[i].first, &logos[i].second);
```
Comment (SansPapyrus683) [inline solutions/bronze/cf-581D.mdx:28]: why scanf/printf now

File: solutions/bronze/cf-581D.mdx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+								goto outer;
+							}
+							grid[r][c] = num_placed + 'A';
+						}
+					}
+					num_placed++;
+				}
+			}
+		}
+		assert(num_placed == 3);
```
Comment (SansPapyrus683) [inline solutions/bronze/cf-581D.mdx:71]: explain this one

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/bronze/cf-581D.mdx (hunk 2 above)
Comment (envyaims) [reply solutions/bronze/cf-581D.mdx:46]: huh?

File: solutions/bronze/cf-581D.mdx (hunk 4 above)
Comment (envyaims) [reply solutions/bronze/cf-581D.mdx:71]: explain what? you have 3 logos and a fixed area so all of them have to be placed...

File: solutions/bronze/cf-581D.mdx (hunk 3 above)
Comment (envyaims) [reply solutions/bronze/cf-581D.mdx:28]: yes

File: solutions/bronze/cf-581D.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/bronze/cf-581D.mdx:46]: space relevant code blocks out
my eyes are burning

File: solutions/bronze/cf-581D.mdx (hunk 4 above)
Comment (SansPapyrus683) [reply solutions/bronze/cf-581D.mdx:77]: "make sure all 3 logos are placed"

File: solutions/bronze/cf-581D.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply solutions/bronze/cf-581D.mdx:28]: what

File: solutions/bronze/cf-581D.mdx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+author: Chongtian Ma
+---
+
+[Official Editorial](https://codeforces.com/blog/entry/20508)
+
+## Implementation
+
+We can brute force all rotations of the three rectangles with a bitmask of $2^3$, with bit $i$ on representing the $i$'th rectangle should be rotated by
+$90$ degrees. The rest is just checking if the configurations are valid.
+
```
Comment (ryanchou-dev) [inline solutions/bronze/cf-581D.mdx:14]:
```suggestion
We can brute force all rotations of the three rectangles with a bitmask of $2^3$, with bit $i$ representing whether the $i$'th rectangle should be rotated by
$90$ degrees. The rest is just checking if the configurations are valid.

```

File: solutions/bronze/cf-581D.mdx [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+**Time Complexity:** $\mathcal{O}(2^n \cdot \textbf{side}^4)$, where $n$ is the number of rectangles (in this case, 3) and $\textbf{side}$ as the side length of the square.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+const int n = 3;
```
Comment (ryanchou-dev) [inline solutions/bronze/cf-581D.mdx:23]: capitalize constants

File: solutions/bronze/cf-581D.mdx [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	long long area = 0;
+	for (pair<int, int> p : logos) { area += p.first * p.second; }
+	// if area is not a perfect square, then it's already invalid
+	int len = 1;
+	while (len * len < area) { len++; }
+	if (len * len != area) {
+		printf("-1");
+		return 0;
+	}
```
Comment (ryanchou-dev) [inline solutions/bronze/cf-581D.mdx:38]:
```suggestion
	long long area = 0;
	for (pair<int, int> p : logos) { area += p.first * p.second; }

	// if area is not a perfect square, then it's already invalid
	int len = 1;
	while (len * len < area) { len++; }

	if (len * len != area) {
		printf("-1");
		return 0;
	}
```

**Review by ryanchou-dev — COMMENTED** [review]
can u switch back to cin/cout uwu

File: solutions/bronze/cf-581D.mdx (hunk 4 above)
Comment (envyaims) [reply solutions/bronze/cf-581D.mdx:79]: i added

File: solutions/bronze/cf-581D.mdx (hunk 3 above)
Comment (envyaims) [reply solutions/bronze/cf-581D.mdx:28]: what

Comment (envyaims) [reply solutions/bronze/cf-581D.mdx:28]: does it make a differnce

**Review by jessechoe10 — APPROVED** [review]
The explanation seems good to me :pray:

File: solutions/bronze/cf-581D.mdx (hunk 3 above)
Comment (ryanchou-dev) [reply solutions/bronze/cf-581D.mdx:28]: idk we use cin/cout like everywhere lol

**Review by ryanchou-dev — APPROVED** [review]
js the cin/cout seems weird but the rest looks good 👍

File: solutions/bronze/cf-581D.mdx (hunk 3 above)
Comment (envyaims) [reply solutions/bronze/cf-581D.mdx:28]: scanf users are underrepresented

Comment (SansPapyrus683) [reply solutions/bronze/cf-581D.mdx:28]: for consistency's sake, i'd rather you change it to cin/cout

File: solutions/bronze/cf-581D.mdx [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+		}
+
+		// at this point, all logos must be placed
+		assert(num_placed == N);
+
+		printf("%d\n", len);
+		for (int i = 0; i < len; i++) { printf("%s\n", grid[i].c_str()); }
+		return 0;
+		// continue the iteration
+	outer:;
```
Comment (SansPapyrus683) [inline solutions/bronze/cf-581D.mdx:85]: formatting kinda sus

Comment (envyaims) [reply solutions/bronze/cf-581D.mdx:85]: THE BOT AUTOFORMAT

File: solutions/bronze/cf-581D.mdx (hunk 3 above)
Comment (envyaims) [reply solutions/bronze/cf-581D.mdx:28]: ive used scanf in other places too though

**Pull request merged**